File: packages/tiny-mcp-client-rust/src/layer.js
import { createRequire } from "node:module";
import { McpError } from "./index.js";
const { NativeMessageLayer, NativeRetryState } = createRequire(import.meta.url)(
  "./tiny-mcp-client-rust.node"
);

function unwrap(value) {
  if (value.error !== undefined) {
    const { code, message, data } = value.error;
    throw code === undefined ? new Error(message) : new McpError(code, message, data);
  }
  return value;
}

async function* readLines(input) {
  const decoder = new TextDecoder("utf-8", { fatal: true });
  let parts = [];
  let bytes = 0;
  function append(text) {
    bytes += Buffer.byteLength(text, "utf8");
    if (bytes > 16 * 1024 * 1024)
      throw new Error("Stdio input line byte limit exceeded (16777216 bytes)");
    if (text.length) parts.push(text);
  }
  for await (const chunk of input) {
    const decoded =
      chunk instanceof Uint8Array
        ? decoder.decode(chunk, { stream: true })
        : decoder.decode() + String(chunk);
    let start = 0;
    for (let end = decoded.indexOf("\n"); end !== -1; end = decoded.indexOf("\n", start)) {
      append(decoded.slice(start, end));
      const line = parts.join("");
      yield line.endsWith("\r") ? line.slice(0, -1) : line;
      parts = [];
      bytes = 0;
      start = end + 1;
    }
    append(decoded.slice(start));
  }
  append(decoder.decode());
  if (parts.length) {
    const line = parts.join("");
    yield line.endsWith("\r") ? line.slice(0, -1) : line;
  }
}

export class JsonRpcMessageLayer {
  requestMetadata;
  #native;
  #input;
  #output;
  #closedReason;
  #disposed;
  #pending = new Map();
  #exchanges = new Map();
  #incoming = new Map();
  #handlers = new Map();
  #notifications = new Map();
  #inputHandlers = new Map();
  constructor(
    input,
    output,
    requestTimeoutMs = 30000,
    inputClosedReason,
    maxConcurrentRequests = 128
  ) {
    if (!Number.isFinite(requestTimeoutMs) || requestTimeoutMs < 0)
      throw new Error("requestTimeoutMs must be a non-negative finite number");
    this.#native = new NativeMessageLayer(maxConcurrentRequests);
    this.#input = input;
    this.#output = output;
    this.#closedReason = inputClosedReason;
    this.requestTimeoutMs = requestTimeoutMs;
    output.once("error", (reason) => this.dispose(reason));
    output.once("close", () => {
      if (this.#disposed === undefined)
        void this.#streamReason(new Error("JSON-RPC output closed")).then((reason) =>
          this.dispose(reason)
        );
    });
    void this.#consume();
  }
  onRequest(method, handler) {
    this.#handlers.set(method, handler);
    this.#native.registerRequest(method);
  }
  onNotification(method, handler) {
    this.#notifications.set(method, handler);
  }
  onInputRequest(method, handler) {
    this.#inputHandlers.set(method, handler);
  }
  sendNotification(method, params) {
    if (this.#disposed !== undefined) throw this.#disposed;
    this.#output.write(unwrap(this.#native.notification(method, params)).line);
  }
  sendRequest(method, params, options = {}) {
    if (this.#disposed !== undefined) throw this.#disposed;
    const timeoutMs =
      options.timeoutMs === null ? null : (options.timeoutMs ?? this.requestTimeoutMs);
    if (timeoutMs !== null && (!Number.isFinite(timeoutMs) || timeoutMs < 0))
      throw new Error("timeoutMs must be a non-negative finite number");
    const token = this.#native.beginExchange();
    const controller = new AbortController();
    this.#exchanges.set(token, controller);
    const forward = () => controller.abort(options.signal.reason);
    options.signal?.addEventListener("abort", forward, { once: true });
    let id;
    let rejectAbort;
    const aborted = new Promise((_resolve, reject) => {
      rejectAbort = reject;
    });
    const cancel = () => {
      if (id !== undefined) this.cancelRequest(id, controller.signal.reason);
      rejectAbort(controller.signal.reason);
    };
    controller.signal.addEventListener("abort", cancel, { once: true });
    const once = (args) =>
      new Promise((resolve, reject) => {
        try {
          const prepared = unwrap(
            this.#native.prepareRequest(token, method, args, this.requestMetadata)
          );
          id = prepared.id;
          options.onRequestId?.(id);
          if (options.signal?.aborted) forward();
          controller.signal.throwIfAborted();
          const timeout =
            timeoutMs === null
              ? undefined
              : setTimeout(() => {
                  this.#native.cancelRequest(id);
                  this.#pending.delete(id);
                  try {
                    options.onTimeout?.(id);
                  } finally {
                    reject(
                      new Error(`JSON-RPC request "${method}" timed out after ${timeoutMs}ms`)
                    );
                  }
                }, timeoutMs);
          this.#pending.set(id, { resolve, reject, timeout });
          try {
            this.#output.write(prepared.line);
          } catch (error) {
            this.cancelRequest(id, error);
          }
        } catch (error) {
          reject(error);
        }
      });
    const work = (async () => {
      let retry;
      if (this.requestMetadata !== undefined) {
        try {
          retry = new NativeRetryState(method, params);
        } catch (error) {
          throw new McpError(-32602, error.message);
        }
      }
      let args = params;
      if (options.signal?.aborted) forward();
      while (true) {
        if (this.#disposed !== undefined) throw this.#disposed;
        controller.signal.throwIfAborted();
        const result = await once(args);
        if (this.requestMetadata === undefined) return result;
        if (retry === undefined) retry = new NativeRetryState(method, params);
        const action = unwrap(
          retry.processResult(result, this.requestMetadata, [...this.#inputHandlers.keys()])
        );
        if (action.type === "complete") return action.result;
        for (const request of action.requests) {
          if (this.#disposed !== undefined) throw this.#disposed;
          controller.signal.throwIfAborted();
          const response = await this.#inputHandlers.get(request.method)(request.params, {
            id: request.key,
            method: request.method,
            signal: controller.signal
          });
          unwrap(retry.recordResponse(request.key, request.method, response));
        }
        if (this.#disposed !== undefined) throw this.#disposed;
        controller.signal.throwIfAborted();
        args = retry.nextParams();
      }
    })();
    if (options.signal?.aborted) forward();
    return Promise.race([work, aborted]).finally(() => {
      options.signal?.removeEventListener("abort", forward);
      controller.signal.removeEventListener("abort", cancel);
      this.#exchanges.delete(token);
      this.#native.finishExchange(token);
    });
  }
  cancelRequest(id, reason) {
    const pending = this.#pending.get(id);
    if (pending === undefined) return false;
    const canceled = this.#native.cancelRequest(id);
    if (!canceled) return false;
    this.#pending.delete(id);
    if (pending !== undefined) {
      clearTimeout(pending.timeout);
      pending.reject(reason);
    }
    return true;
  }
  dispose(reason = new Error("JSON-RPC message layer disposed")) {
    if (this.#disposed !== undefined) return;
    this.#disposed = reason;
    for (const controller of this.#exchanges.values()) controller.abort(reason);
    for (const pending of this.#pending.values()) {
      clearTimeout(pending.timeout);
      pending.reject(reason);
    }
    this.#pending.clear();
    for (const controller of this.#incoming.values()) controller.abort(reason);
    this.#incoming.clear();
    this.#native.dispose();
  }
  async #streamReason(fallback = new Error("JSON-RPC input stream closed")) {
    if (this.#closedReason === undefined) return fallback;
    let timeout;
    try {
      return await Promise.race([
        this.#closedReason,
        new Promise((resolve) => {
          timeout = setTimeout(() => resolve(fallback), 50);
        })
      ]);
    } catch {
      return fallback;
    } finally {
      clearTimeout(timeout);
    }
  }
  async #consume() {
    try {
      for await (const line of readLines(this.#input)) {
        if (this.#disposed !== undefined) break;
        this.#native.feed(line);
        while (this.#disposed === undefined) {
          this.#native.setModern(this.requestMetadata !== undefined);
          const event = this.#native.nextEvent();
          if (event === null || event === undefined) break;
          if (event.type === "write") this.#output.write(event.line);
          else if (event.type === "settle") {
            const pending = this.#pending.get(event.id);
            this.#pending.delete(event.id);
            if (pending === undefined) continue;
            clearTimeout(pending.timeout);
            if (Object.hasOwn(event, "result")) pending.resolve(event.result);
            else
              pending.reject(new McpError(event.error.code, event.error.message, event.error.data));
          } else if (event.type === "cancel")
            this.#incoming.get(event.token)?.abort(new Error(event.reason));
          else if (event.type === "notification") {
            try {
              await this.#notifications.get(event.method)?.(event.params, { method: event.method });
            } catch {
              // Observer failures do not terminate message input.
            }
          } else if (event.type === "invoke")
            void this.#invoke(event).catch((error) => this.dispose(error));
        }
      }
      if (this.#disposed === undefined) this.dispose(await this.#streamReason());
    } catch (error) {
      this.dispose(
        error instanceof Error ? error : new Error(`JSON-RPC input stream failed: ${String(error)}`)
      );
    }
  }
  async #invoke(event) {
    const controller = new AbortController();
    this.#incoming.set(event.token, controller);
    try {
      const invalid = this.#native.validateIncoming(event.token, event.params);
      if (invalid !== undefined && invalid !== null) {
        this.#output.write(invalid);
        return;
      }
      const result = await this.#handlers.get(event.method)(event.params, {
        id: event.id,
        method: event.method,
        signal: controller.signal
      });
      if (this.#disposed !== undefined || controller.signal.aborted) return;
      const line = this.#native.completeIncoming(event.token, result, false);
      if (line !== undefined && line !== null) this.#output.write(line);
    } catch (error) {
      if (this.#disposed !== undefined || controller.signal.aborted) return;
      const message = error instanceof Error ? error.message : String(error);
      const line = this.#native.completeIncoming(
        event.token,
        { code: error instanceof McpError ? error.code : -32603, message },
        true
      );
      if (line !== undefined && line !== null) this.#output.write(line);
    } finally {
      this.#incoming.delete(event.token);
      this.#native.finishIncoming(event.token);
    }
  }
}

File: packages/tiny-mcp-client-rust/src/layer.rs
//! Protocol bookkeeping is independent of streams, callbacks and clock adapters.
use crate::messages::{ParsedMessage, parse_payload};
use mcp_protocol_rust::{
    json::{self, Limits, Value},
    metadata::is_valid_metadata,
};
use std::collections::{HashMap, HashSet, VecDeque};
use tiny_stdio_mcp_server_rust::protocol::validate_definition;

#[derive(Debug, PartialEq)]
pub struct LayerError {
    pub code: Option<i32>,
    pub message: &'static str,
}
#[derive(Debug, PartialEq)]
pub enum Event {
    Write(Value),
    Settle {
        id: u64,
        result: Option<Value>,
        error: Option<Value>,
    },
    Invoke {
        token: u64,
        id: Value,
        method: Vec<u16>,
        params: Option<Value>,
    },
    Cancel {
        token: u64,
        reason: Vec<u16>,
    },
    Notification {
        method: Vec<u16>,
        params: Option<Value>,
    },
}
#[derive(Hash, PartialEq, Eq)]
enum Key {
    String(Vec<u16>),
    Number(u64),
}
impl Key {
    fn from_value(value: &Value) -> Option<Self> {
        match value {
            Value::String(text) => Some(Self::String(text.clone())),
            Value::Number(number) => Some(Self::Number(if *number == 0.0 {
                0
            } else if number.is_nan() {
                f64::NAN.to_bits()
            } else {
                number.to_bits()
            })),
            _ => None,
        }
    }
}
struct Incoming {
    id: Value,
    method: Vec<u16>,
    canceled: bool,
}
pub struct MessageLayer {
    limit: usize,
    modern: bool,
    disposed: bool,
    next_exchange: u64,
    next_request: u64,
    next_incoming: u64,
    exchanges: HashSet<u64>,
    pending: HashMap<u64, u64>,
    handlers: HashSet<Vec<u16>>,
    incoming: HashMap<u64, Incoming>,
    named_incoming: HashMap<Key, u64>,
    queued: VecDeque<ParsedMessage>,
    events: VecDeque<Event>,
}
impl MessageLayer {
    pub fn new(limit: usize) -> Result<Self, LayerError> {
        if limit == 0 {
            return Err(error(
                None,
                "maxConcurrentRequests must be a positive safe integer",
            ));
        }
        Ok(Self {
            limit,
            modern: false,
            disposed: false,
            next_exchange: 0,
            next_request: 0,
            next_incoming: 0,
            exchanges: HashSet::new(),
            pending: HashMap::new(),
            handlers: HashSet::new(),
            incoming: HashMap::new(),
            named_incoming: HashMap::new(),
            queued: VecDeque::new(),
            events: VecDeque::new(),
        })
    }
    pub fn set_modern(&mut self, modern: bool) {
        self.modern = modern;
    }
    pub fn register_request(&mut self, method: Vec<u16>) {
        self.handlers.insert(method);
    }
    pub fn begin_exchange(&mut self) -> Result<u64, LayerError> {
        if self.disposed {
            return Err(error(None, "JSON-RPC message layer disposed"));
        }
        if self.exchanges.len() >= self.limit {
            return Err(error(None, "JSON-RPC request capacity exceeded"));
        }
        self.next_exchange = next(self.next_exchange)?;
        self.exchanges.insert(self.next_exchange);
        Ok(self.next_exchange)
    }
    pub fn prepare_request(
        &mut self,
        exchange: u64,
        method: Vec<u16>,
        params: Option<Value>,
        metadata: Option<Value>,
    ) -> Result<Value, LayerError> {
        if self.disposed {
            return Err(error(None, "JSON-RPC message layer disposed"));
        }
        if !self.exchanges.contains(&exchange) {
            return Err(error(None, "Unknown JSON-RPC exchange"));
        }
        if params
            .as_ref()
            .is_some_and(|params| !params.is_json_value())
        {
            return Err(error(
                Some(-32602),
                "Request params must contain only JSON values",
            ));
        }
        if !valid_params_metadata(params.as_ref())
            || metadata
                .as_ref()
                .is_some_and(|metadata| !is_valid_metadata(metadata))
        {
            return Err(error(Some(-32602), "Invalid request metadata"));
        }
        self.next_request = next(self.next_request)?;
        let params = if let Some(Value::Object(metadata)) = metadata {
            let mut params = match params {
                None => vec![],
                Some(Value::Object(params)) => params,
                _ => {
                    return Err(error(
                        Some(-32602),
                        "Modern request params must be an object",
                    ));
                }
            };
            let mut merged = params
                .iter()
                .find(|(key, _)| key == &units("_meta"))
                .and_then(|(_, value)| match value {
                    Value::Object(entries) => Some(entries.clone()),
                    _ => None,
                })
                .unwrap_or_default();
            for (key, value) in metadata {
                put(&mut merged, key, value);
            }
            put(&mut params, units("_meta"), Value::Object(merged));
            Some(Value::Object(params))
        } else {
            params
        };
        let id = self.next_request;
        let mut request = vec![
            (units("jsonrpc"), Value::String(units("2.0"))),
            (units("id"), Value::Number(id as f64)),
            (units("method"), Value::String(method)),
        ];
        if let Some(params) = params {
            request.push((units("params"), params));
        }
        self.pending.insert(id, exchange);
        Ok(Value::Object(request))
    }
    pub fn notification(
        &self,
        method: Vec<u16>,
        params: Option<Value>,
    ) -> Result<Value, LayerError> {
        if self.disposed {
            return Err(error(None, "JSON-RPC message layer disposed"));
        }
        if params
            .as_ref()
            .is_some_and(|params| !params.is_json_value())
        {
            return Err(error(
                Some(-32602),
                "Notification params must contain only JSON values",
            ));
        }
        if !valid_params_metadata(params.as_ref()) {
            return Err(error(Some(-32602), "Invalid notification metadata"));
        }
        let mut result = vec![
            (units("jsonrpc"), Value::String(units("2.0"))),
            (units("method"), Value::String(method)),
        ];
        if let Some(params) = params {
            result.push((units("params"), params));
        }
        Ok(Value::Object(result))
    }
    pub fn cancel_request(&mut self, id: u64) -> bool {
        self.pending.remove(&id).is_some()
    }
    pub fn finish_exchange(&mut self, exchange: u64) -> bool {
        self.pending.retain(|_, owner| *owner != exchange);
        self.exchanges.remove(&exchange)
    }
    pub fn feed(&mut self, line: &[u16]) -> Result<(), LayerError> {
        if self.disposed || line.is_empty() {
            return Ok(());
        }
        match json::parse_utf16(line, Limits::default()) {
            Ok(Value::Array(batch)) => self.queued.extend(batch.into_iter().map(parse_payload)),
            Ok(value) => self.queued.push_back(parse_payload(value)),
            Err(_) => self.queued.push_back(ParsedMessage::Invalid {
                id: Value::Null,
                code: -32700,
                message: "Parse error",
            }),
        }
        Ok(())
    }
    pub fn next_event(&mut self) -> Option<Event> {
        if self.disposed {
            return None;
        }
        if let Some(event) = self.events.pop_front() {
            return Some(event);
        }
        while let Some(parsed) = self.queued.pop_front() {
            match parsed {
                ParsedMessage::Invalid { id, code, message } => {
                    return Some(Event::Write(error_reply(id, code as f64, units(message))));
                }
                ParsedMessage::Response(message) => {
                    let Some(Value::Number(id)) = message.get("id") else {
                        continue;
                    };
                    if !id.is_finite()
                        || *id < 1.0
                        || id.fract() != 0.0
                        || !self.pending.contains_key(&(*id as u64))
                    {
                        continue;
                    }
                    let id = *id as u64;
                    self.pending.remove(&id);
                    return Some(Event::Settle {
                        id,
                        result: message.get("result").cloned(),
                        error: message.get("error").cloned(),
                    });
                }
                ParsedMessage::Notification(message) => {
                    let Value::String(method) =
                        message.get("method").expect("parsed notification method")
                    else {
                        unreachable!()
                    };
                    let params = message.get("params").cloned();
                    let notification = Event::Notification {
                        method: method.clone(),
                        params: params.clone(),
                    };
                    if method == &units("notifications/cancelled")
                        && let Some(params) = &params
                        && let Some(key) = params.get("requestId").and_then(Key::from_value)
                        && let Some(token) = self.named_incoming.get(&key)
                        && let Some(incoming) = self.incoming.get_mut(token)
                    {
                        incoming.canceled = true;
                        let reason = match params.get("reason") {
                            Some(Value::String(reason)) => reason.clone(),
                            _ => units("Server request cancelled"),
                        };
                        self.events.push_back(notification);
                        return Some(Event::Cancel {
                            token: *token,
                            reason,
                        });
                    }
                    return Some(notification);
                }
                ParsedMessage::Request(message) => {
                    let id = message.get("id").expect("parsed request id").clone();
                    let Value::String(method) =
                        message.get("method").expect("parsed request method")
                    else {
                        unreachable!()
                    };
                    if self.modern {
                        return Some(Event::Write(error_reply(
                            id,
                            -32600.0,
                            units("Modern MCP servers cannot initiate JSON-RPC requests"),
                        )));
                    }
                    if !self.handlers.contains(method) {
                        return Some(Event::Write(error_reply(
                            id,
                            -32601.0,
                            units("Method not found: ")
                                .into_iter()
                                .chain(method.iter().copied())
                                .collect(),
                        )));
                    }
                    let key = Key::from_value(&id).expect("parsed request key");
                    if self.named_incoming.contains_key(&key) {
                        return Some(Event::Write(error_reply(
                            id,
                            -32600.0,
                            units("Duplicate active server request ID"),
                        )));
                    }
                    if self.incoming.len() >= self.limit {
                        return Some(Event::Write(error_reply(
                            id,
                            -32000.0,
                            units("Server request capacity exceeded"),
                        )));
                    }
                    let token = match next(self.next_incoming) {
                        Ok(token) => token,
                        Err(_) => {
                            return Some(Event::Write(error_reply(
                                id,
                                -32000.0,
                                units("Server request identifier exhausted"),
                            )));
                        }
                    };
                    self.next_incoming = token;
                    self.named_incoming.insert(key, token);
                    self.incoming.insert(
                        token,
                        Incoming {
                            id: id.clone(),
                            method: method.clone(),
                            canceled: false,
                        },
                    );
                    return Some(Event::Invoke {
                        token,
                        id,
                        method: method.clone(),
                        params: message.get("params").cloned(),
                    });
                }
            }
        }
        None
    }
    pub fn validate_incoming(&self, token: u64, params: Option<Value>) -> Option<Value> {
        let incoming = self.incoming.get(&token)?;
        response_type(&incoming.method)?;
        let mut input = vec![(units("method"), Value::String(incoming.method.clone()))];
        if let Some(params) = params {
            input.push((units("params"), params));
        }
        if validate_definition("InputRequest", &Value::Object(input)) {
            return None;
        }
        Some(error_reply(
            incoming.id.clone(),
            -32602.0,
            units("Invalid client input request"),
        ))
    }
    pub fn complete_incoming(&self, token: u64, result: Result<Value, Value>) -> Option<Value> {
        let incoming = self.incoming.get(&token)?;
        if self.disposed || incoming.canceled {
            return None;
        }
        let (key, result) = match result {
            Ok(result) => {
                if response_type(&incoming.method)
                    .is_some_and(|definition| !validate_definition(definition, &result))
                {
                    return Some(error_reply(
                        incoming.id.clone(),
                        -32603.0,
                        units("Invalid client input response"),
                    ));
                }
                if !result.is_json_value() {
                    return Some(error_reply(
                        incoming.id.clone(),
                        -32603.0,
                        units("Response result must contain only JSON values"),
                    ));
                }
                ("result", result)
            }
            Err(error) => ("error", error),
        };
        Some(Value::Object(vec![
            (units("jsonrpc"), Value::String(units("2.0"))),
            (units("id"), incoming.id.clone()),
            (units(key), result),
        ]))
    }
    pub fn finish_incoming(&mut self, token: u64) -> bool {
        let Some(incoming) = self.incoming.remove(&token) else {
            return false;
        };
        if let Some(key) = Key::from_value(&incoming.id) {
            self.named_incoming.remove(&key);
        }
        true
    }
    pub fn dispose(&mut self) {
        self.disposed = true;
        self.pending.clear();
        self.exchanges.clear();
        self.incoming.clear();
        self.named_incoming.clear();
        self.queued.clear();
        self.events.clear();
    }
}
fn next(value: u64) -> Result<u64, LayerError> {
    value
        .checked_add(1)
        .filter(|value| *value <= 9_007_199_254_740_991)
        .ok_or_else(|| error(None, "JSON-RPC request identifier exhausted"))
}
fn units(value: &str) -> Vec<u16> {
    value.encode_utf16().collect()
}
fn error(code: Option<i32>, message: &'static str) -> LayerError {
    LayerError { code, message }
}
fn put(entries: &mut Vec<(Vec<u16>, Value)>, key: Vec<u16>, value: Value) {
    if let Some((_, previous)) = entries.iter_mut().find(|(name, _)| name == &key) {
        *previous = value;
    } else {
        entries.push((key, value));
    }
}
fn valid_params_metadata(params: Option<&Value>) -> bool {
    params
        .and_then(|params| params.get("_meta"))
        .is_none_or(is_valid_metadata)
}
fn response_type(method: &[u16]) -> Option<&'static str> {
    match String::from_utf16_lossy(method).as_str() {
        "roots/list" => Some("ListRootsResult"),
        "sampling/createMessage" => Some("CreateMessageResult"),
        "elicitation/create" => Some("ElicitResult"),
        _ => None,
    }
}
fn error_reply(id: Value, code: f64, message: Vec<u16>) -> Value {
    Value::Object(vec![
        (units("jsonrpc"), Value::String(units("2.0"))),
        (units("id"), id),
        (
            units("error"),
            Value::Object(vec![
                (units("code"), Value::Number(code)),
                (units("message"), Value::String(message)),
            ]),
        ),
    ])
}
